File: .old/Aclass_one_more_time/A1107.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
int father[1010];
int isRoot[1010];
int hobby[1010];

int find_father(int son){
    if(father[son] != son){
        father[son] = find_father(father[son]);
    }
    return father[son];
}

int main(){
    int n;
    cin >> n;
    for(int i = 1; i <= n; i++){
        father[i] = i;
    }
    for(int i = 1; i <= n; i++){
        int temp_num;
        scanf("%d: ", &temp_num);
        for(int j = 0; j < temp_num; j++){
            int temp;
            cin >> temp;
            if(hobby[temp] == 0) hobby[temp] = i;
            else{
                father[find_father(i)] = find_father(hobby[temp]);
            }
        }
    }
    for(int i = 1; i <= n; i++){
        isRoot[find_father(i)] += 1;
    }
    vector<int> ans;
    for(int i = 1; i <= n; i++){
        if(isRoot[i] != 0) ans.push_back(isRoot[i]);
    }
    sort(ans.begin(), ans.end());
    reverse(ans.begin(), ans.end());
    cout << ans.size() << endl;
    for(int i = 0; i < ans.size(); i++){
        if(i != 0) cout << " " << ans[i];
        else cout << ans[i];
    }
    cout << endl;
    return 0;
}

File: .old/Aclass_Python/A1107.py
n = int(input())
hobby = [0 for i in range(1001)]
father = [i for i in range(n + 1)]
isRoot = [0 for i in range(n + 1)]

def findFather(son):
    if son != father[son]:
        father[son] = findFather(father[son])
    return father[son]

for i in range(n):
    hobbies = list(map(int, input().split()[1:]))
    for j in hobbies:
        if hobby[j] == 0:
            hobby[j] = i + 1
        else:
            father[findFather(i + 1)] = findFather(hobby[j])
            

for i in range(n):
    isRoot[findFather(i + 1)] += 1
ans = []
for i in isRoot:
    if i != 0:
        ans.append(i)
print(len(ans))
print(" ".join(list(map(str, sorted(ans, reverse=True)))))